### DISCUSSION TITLE
Integration with firebase functions?

### DISCUSSION THREAD

**Discussion by mfbx9da4** [trpc/trpc#1263] [Q&A]
Is there an adapter for firebase functions. Is there any guidance for writing custom adapters?

**Comment by KATT** [comment]
Haven't thought about exactly that'd would work but I can't think of any blockers on why firebase fns wouldn't work.

Clone the repo, have a look in packages/server/src/adapters for inspo and potentially also the fastify test

**Comment by camelmasa** [comment]
We could  do that using Express adapter.
Ref: https://trpc.io/docs/express

**Reply by KATT** [reply]
Would be great if someone did this, but we won't add any more adapters to trpc for the time being.

We're happy to provide assistance though & help promoting a library that did a nice firebase integration

**Comment by WesleyYue** [comment]
Thanks @camelmasa for pointing out the Express adapter. You can already host all the trpc routes as an entire Express app in one Firebase function. This is how I did it. The downsides compared to vanilla Firebase functions:
- No support for non-http event triggers (like pubsub, firestore triggers)
    - You can probably create a shim with a firebase function that has those triggers and then do server-side calls. I haven't tried this yet so unsure what pitfalls might come up.
- Stuffing large bundle to one Firebase function instance. Potentially longer cold start times (but probably not as big of an issue with GCF 2nd gen)
- GCP logs will all fall under the same function, so you can't use `resource.labels.function_name` filters



```
// server.ts
import type * as trpcExpress from "@trpc/server/adapters/express";
import type { inferAsyncReturnType } from "@trpc/server";
import { initTRPC } from "@trpc/server";
import { z } from "zod";

export const createContext = ({
  req,
  res,
}: trpcExpress.CreateExpressContextOptions): { name: string } => ({
  name: "test",
}); // no context
type Context = inferAsyncReturnType<typeof createContext>;

const t = initTRPC.context<Context>().create();
export const router = t.router({
  asyncTest: t.procedure.query(() => {
    // Functions that return a Promise will not terminate until the Promise is resolved, just like regular Firebase Functions
    return new Promise((resolve) => {
      setTimeout(() => {
        resolve("Hello, world!");
      }, 10000);
    });
  }),
});

export type AppRouter = typeof router;
```

```
// trpc.ts
import * as functions from "firebase-functions";
import * as trpcExpress from "@trpc/server/adapters/express";
import express from "express";
import cors from "cors";
import { createContext, router } from "./server";

const app = express();

const corsOrigins = ["https://example.com"];

if (process.env.FUNCTIONS_EMULATOR === "true") {
  corsOrigins.push("http://localhost:3000");
}

app.use(
  cors({
    origin: corsOrigins,
    credentials: true,
  })
);
app.use(
  // Express app already mounts to the "trpc" Firebase function. So Express/trpc routes will be at
  // http(s)://<firebase-host>/<project-id>/<region>/trpc/<route-name>. Firebase routes trpc/**
  // to the "trpc" function via firebase.json.
  "/",
  trpcExpress.createExpressMiddleware({
    router,
    createContext,
  })
);

export const trpc = functions.region("us-west1").https.onRequest(app);
```

```
// firebase.json
{
  ...
  "hosting": {
    "rewrites": [
      {
        "source": "/trpc/**",
        "function": "trpc"
      }
    ]
  }
}
```

**Comment by yairopro** [comment]
##  Without express ☕️

```typescript
import { onRequest } from 'firebase-functions/v2/https';
import { createHTTPHandler } from '@trpc/server/adapters/standalone';

const router = trpc.router({
  // ...
});


export const myFireFunction = onRequest(createHTTPHandler({ router }));
```

**Comment by xeno-sun** [comment]
A minimalist demo:

```ts
import { initTRPC } from "@trpc/server";
import { createHTTPHandler } from "@trpc/server/adapters/standalone";
import { onRequest } from "firebase-functions/v1/https";

const t = initTRPC.create();

const appRouter = t.router({
  Hi: t.procedure.query(() => {
    return { data: "hi" };
  }),
});

export const v0 = onRequest(
  createHTTPHandler({
    router: appRouter,
    createContext: () => {
      return {};
    },
  }),
);
```

### Warning

> Failed to load function definition from source: FirebaseError: Functions codebase could not be analyzed successfully. It may have a syntax or runtime error

If you encountered the error mentioned above, which is thrown by firebase-tools, I recommend reviewing you Firebase functions source files. Ensure that you don't export any variables or data using the "export" statement, except for the APIs you want to export.

**Reply by yairopro** [reply]
For me, to find the origin of my `FirebaseError: Functions codebase could not be analyzed successfully` I tried to run the transpiled code (in `dist/`) with `node` directly. It outputs in the terminal the original errors which are much more verbose.

**Comment by ryan-saffer** [comment]
I have been using tRPC with firebase functions that hosts different routes on separate functions. This enables me to continue to separate my backend into discrete, independently scalable & deployable 'microservices', each with their own set of procedures, while on the client it looks and feels like a traditional server and tRPC setup. Here is my approach:
```typescript
// functions/trpc.ts

import type { NodeHTTPCreateContextFnOptions } from '@trpc/server/dist/adapters/node-http'
import type { IncomingMessage, ServerResponse } from 'http'
import { createHTTPHandler } from '@trpc/server/adapters/standalone'
// .. other imports left out for brevity

// INITIALISATION
const t = initTRPC.context<typeof createContext>().create()
const router = t.router
const middleware = t.middleware

// CONTEXT
export function createContext({
    req,
}: NodeHTTPCreateContextFnOptions<IncomingMessage, ServerResponse<IncomingMessage>>) {
    return { authToken: req.headers.authorization }
}

// MIDDLEWARE
const isAuthenticated = middleware(async ({ ctx, next }) => {
    try {
        await getAuth().verifyIdToken(ctx.authToken || '') // https://firebase.google.com/docs/auth/admin/verify-id-tokens
        return next()
    } catch {
        throw new HttpsError('unauthenticated', 'procedure requires authentication')
    }
})

// PROCEDURES
const publicProcedure = t.procedure
const authenticatedProcedure = publicProcedure.use(isAuthenticated)

// ROUTERS
const firstRouter = router({
    firstRouterQuery: publicProcedure.query(() => {
        return { result: `first router query result` }
    }),
})

const secondRouter = router({
    secondRouterMutation: authenticatedProcedure
        .input(
            z.object({
                foo: z.string(),
            })
        )
        .mutation((opts) => {
            return { result: `data sent was ${opts.input.foo}` }
        }),
})

// APP ROUTER - only created to export type
const appRouter = router({
    firstFirebaseFunction: firstRouter,
    secondFirebaseFunction: secondRouter
})
export type AppRouter = typeof appRouter
```

```typescript
// functions/index.ts

export const firstFirebaseFunction = onRequest(
    { cors: true },
    createHTTPHandler({
        router: firstRouter,
        createContext,
    })
)

export const secondFirebaseFunction = onRequest(
    { cors: true },
    createHTTPHandler({
        router: secondRouter,
        createContext,
    })
)
```
Then in my client, I redirect the trpc procedure to the appropriate cloud function. It's important to note that the reason this works, is because the name of the firebase function (`firstFirebaseFunction` and `secondFirebaseFunction` in this case) matches exactly to their keys in the `appRouter`. My client uses the React Query Integration, but of course it can be adapted for any client:

```typescript
// client/App.tsx

export const App = () => {
    const firebase = useFirebase() // custom hook to get my firebase instance
    const [queryClient] = useState(() => new QueryClient())
    const [trpcClient] = useState(() =>
        trpc.createClient({
            links: [
                httpLink({
                    url: '',
                    async headers() {
                        const token = (await firebase.auth.currentUser?.getIdToken()) || ''
                        return {
                            authorization: token,
                        }
                    },
                    fetch(url, options) {
                        // trpc sends the url in the form '/{router}.{procedure}?{queryParams}'. we just need to replace the first '.' with '/'
                        // note that if you want nested routes within a single firebase function, this will need tweaking.
                        const normalisedUrl = url.toString().replace('.', '/') // only the first instance
                        const splitUrl = normalisedUrl.split('/')
                        const [router, procedure] = [splitUrl[1], splitUrl[2]]
                        const domain = // base firebase functions url
                        return fetch(`${domain}/${router}/${procedure}`, options)
                    },
                }),
            ],
        })
    )

    return (
        <trpc.Provider client={trpcClient} queryClient={queryClient}>
            <QueryClientProvider client={queryClient}>
                {/* Your app here... */}
            </QueryClientProvider>
        </trpc.Provider>
    )
}
```
```typescript
// client/trpc.ts

import type { AppRouter } from '../functions/src/trpc/trpc.app-router'
import { createTRPCReact } from '@trpc/react-query'

export const trpc = createTRPCReact<AppRouter>()
```
Now when I use the trpc client anywhere in my app, everything works beautifully, and I can still keep different parts of my backend separated into their own firebase functions, each with their own set of procedures!

**Reply by fuelkoy** [reply]
Hello @ryan-saffer. Your solution is fantastic! While looking at your solution I was wondering where does NodeHTTPCreateContextFnOptions come from?

**Reply by ryan-saffer** [reply]
@fuelkoy imports look like this:
```typescript
// functions/trpc.ts
import type { NodeHTTPCreateContextFnOptions } from '@trpc/server/dist/adapters/node-http'
import type { IncomingMessage, ServerResponse } from 'http'

// ...
```
I was able to solve this part by looking at the type of the options object that the `createHTTPHandler()` function from tRPC accepts. Digging into it you can find this type:
```typescript
// @trpc/server/src/adapters/node-http/types.ts

export type NodeHTTPCreateContextOption<
  TRouter extends AnyRouter,
  TRequest,
  TResponse,
> = object extends inferRouterContext<TRouter>
  ? {
      /**
       * @link https://trpc.io/docs/context
       **/
      createContext?: NodeHTTPCreateContextFn<TRouter, TRequest, TResponse>; // >> This is the type we need! You can go one level deeper to find the `NodeHTTPCreateContextFnOptions<TRequest, TResponse>` which I am using
    }
  : {
      /**
       * @link https://trpc.io/docs/context
       **/
      createContext: NodeHTTPCreateContextFn<TRouter, TRequest, TResponse>;
    };
```
`TRequest` and `TResponse` just needed to then be typed as `IncomingMessage` and `ServerResponse<IncomingMessage>`, which was already inferred by tRPC, so I just made my `createContext()` definition match 😄

**Reply by fuelkoy** [reply]
@ryan-saffer Really great! Thank you putting the time to share here. I have been 🚀looking for a solution just like this and now I have landed to a great one🪂 

Also is `createHTTPHandler` from 'import { createHTTPHandler } from '@trpc/server/adapters/standalone';' or is it another adabter?

**Reply by ryan-saffer** [reply]
@fuelkoy happy to hear I helped out. I personally had a huge amount of firebase functions, and while I had my own type safe 'model' for calling them, I couldn't find a way to combine multiple methods within a single firebase function. This solution has significantly reduced the number of functions I have deployed! Plus tRPC has been such a pleasure to work with.

And yes, the adapter is the standalone adapter:
```typescript
import { createHTTPHandler } from '@trpc/server/adapters/standalone'
```

I have updated my original comment with imports so others won't get confused.

**Reply by fuelkoy** [reply]
@ryan-saffer I have been using your solution while learing trpc and react-query, thank you again for it. 1 Thing I need to ask as I haven't got mutations to work 'beautifully' (yet). How do you call in your example secondRouterMutation on client?

**Got it working got it working** -- wrong tanstack/react-queyry pkg version: had latest as trpc ^10 uses still 4.18.0

Yet thank you again for you comments, made this all possible and it is so pleasent to use trpc!

**Comment by EdRW** [comment]
I am super grateful to everyone here for showing how to integrate tRPC with firebase functions using the built-in adapters! 🙏🏿 
To hopefully make things a bit simpler for myself and others in the future, I've publish an npm package using what I learned here.

# [trpc-firebase-functions](https://www.npmjs.com/package/trpc-firebase-functions)

It is basically just a thin wrapper on top of `@trpc/server/adapters/node-http`, but with a couple small additional benefits.
- convenience
- the type of the `req` property matches the [`Request`](https://github.com/firebase/firebase-functions/blob/9c818713db511895a33378859ab1b9f2eef99179/src/common/providers/https.ts#L45) type defined in `firebase-functions/v2/https`. Helpful when when creating the context and in the `onError` function, and you need the `req.rawBody` property.
- the type of the handler function returned by `createFirebaseHandler` exactly matches the type expected by [`onRequest`](https://github.com/firebase/firebase-functions/blob/9c818713db511895a33378859ab1b9f2eef99179/src/v2/providers/https.ts#L242).

Currently it only supports `onRequest` v2 firebase functions and tRPC v10. However, I plan to add support for `onRequest` v1 firebase functions and tRPC v11. I'm also wondering if there may be a benefit to supporting `onCall` functions as well.

If anyone has time, please take a look. I'd really appreciate any suggestions or feedback.

## Usage

Create the context
```typescript

import type { CreateFirebaseContextOptions } from 'trpc-firebase-functions';

// created for each request
export const createContext = async ({
  req,
  res,
}: CreateFirebaseContextOptions) => {
  req.rawBody // Buffer
};

export type Context = Awaited<ReturnType<typeof createContext>>;
```

Initialize tRPC and create router
```typescript
import { initTRPC } from '@trpc/server';

const t = initTRPC.context<Context>().create();

export const appRouter = t.router({
  // [...]
});

export type AppRouter = typeof appRouter;
```

Use the adapter
```typescript
import { createFirebaseHandler } from 'trpc-firebase-functions';
import { onRequest } from 'firebase-functions/v2/https';

const handler = createFirebaseHandler({
  router: appRouter,
  createContext,
});

export const helloWorld = onRequest(handler);
```

**Reply by saenzramiro** [reply]
Hello @EdRW.
I'm new to tRPC, and having support for onCall functions is mandatory for me. Could you make some progress on that?
